packages/core/src/application_init.ts

```diff
@@ -6,11 +6,20 @@
  * found in the LICENSE file at https://angular.io/license
  */
 
+import {Observable} from 'rxjs';
 import {Inject, Injectable, InjectionToken, Optional} from './di';
 import {isObservable, isPromise} from './util/lang';
 import {noop} from './util/noop';
 
 
+
+/**
+ * A type describing {@link APP_INITIALIZER} callback.
+ *
+ * @publicApi
+ */
+export type AppInitializerCallback = () => Observable<unknown>|Promise<unknown>;
```

Comment:
This ought to be 
```suggestion
export type AppInitializerCallback = () => Observable<unknown>|Promise<unknown>|void;
```

shouldn't it?

Reply:
I am not really sure, because if we take a look at 

https://github.com/angular/angular/blob/master/packages/core/src/application_init.ts#L66-L78

We can see that we don't handle a void callback, so I think passing a void callback will never initialize the app, isn't it ? Or am I missing something ?

Reply:
no actually, you are right, we do complete the promise here https://github.com/angular/angular/blob/master/packages/core/src/application_init.ts#L88-L90

Reply:
The code you linked handles two cases explicitly:

1. A `Promise` return type
2. An `Observable` return type

All other types are simply ignored; the initializer will have been executed but the framework won't have to wait for it.

Arguably the return type could be `any` here; consider the following example:

```ts
@NgModule({
  providers: [{
    provide: APP_INITIALIZER,
    multi: true,
    useValue: () => callSomeFunction(),
  }]
})
export class Mod {}

export function someFunction(): string {
  return 'example';
}
```

Here, the arrow function implicitly returns its expression; therefore a `string` is returned from the initializer callback.

Anyway, I think `void` is the better type here as it documents the intent of an initializer: either it returns a completion signal (`Promise` or `Observable`), or nothing of interest to the framework (`void`).

An additional observation is that the type here is _in no way_ type-checked; a provider registration doesn't constrain the value/class/function that is provided for an `InjectionToken` to its generic type arg.

Reply:
Alright I see, I will make it non-public

Reply:
@JoostK given your explanation for using `void`, I guess the Observable and Promise should also be `void` instead of `unknown`?

Reply:
@alfaproject although that could indeed be done, it has the undesirable side-effect that you can't return e.g. an `Observable<string>` as you'd get `TS2322: Type 'Observable<string>' is not assignable to type 'Observable<void>'.` which is rather unhelpful. You'd then have to actively map the Observable into an `undefined` value (e.g. using `mapTo(undefined)`) or cast it to the return type, both of which aren't very ergonomic.

Reply:
But isn't that the same issue with the last type being `void` instead of `unknown`? https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABMOcBGBDATgCgJQBciAbnDACaIDeAsAFCKJYCmUIWSA5JwNz0C+9IA

Reply:
You could argue that, indeed; but if that function is itself meant to be an initializer function then it should just omit the return value in the first place (which is a lot more ergonomic to achieve than needing an explicit `mapTo(undefined)`); and if the function is used indirectly then there won't be an error:

https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABMOcBGBDATgCgJQBciAzlFjGAOaIDeAsAFCKJYCmUIWSA5NwNyMAvo0YQEpZKiL5EAXgB8iAG5wYAEzmT02AQyA

Reply:
Aye makes sense. Cheers!